### ISSUE TITLE
Research Q&A Pairs Pipeline

### ISSUE THREAD

**Issue by grayJiaaoLi** [amosproj/amos2024ss08-cloud-native-llm#27]
## User story
1. As a software architect
2. I want to research the design of the Q&A pairs pipeline
3. So that the processing is reliable and efficient

## Acceptance criteria

* Determine the tools and solutions that can be used to implement the pipeline
    * Input: 
        * align with previous data extraction
    * Process: 
        * Qs extraction
        * As matching
    * Output: 
        * prepare the data for training

* Write brief descriptions of the solution
    * Cite the reference sources e.g. papers, articles,...
    * (Optional) Design an architecture draft of the pipeline
        * NOT necessary considering the possibility of further development

## Further informations
- you are allowed to use tools under a permissive license (you do not need to reinvent the wheel)

## Definition of done (DoD)
* Added only after week 5
* The same for all features
* Here goes the project-specific part

## DoD general criteria
* Feature has been fully implemented
* Feature has been merged into the mainline
* All acceptance criteria were met
* Product owner-approved features
* All tests are passing
* Developers agreed to release

**Comment by dnsch** [comment]
# Report on Q&A Pairs Pipeline Research Findings

We are trying to enhance our pipeline so that we can extract question/answer pairs from the gathered data that can be then used to finetune our LLM model.

## Findings

1. The most promising existing option appears to be the existing lm-question-generation package https://github.com/asahi417/lm-question-generation?tab=readme-ov-file. It uses the t5 QG model for automatic question/answer or question-answer generation given input data. For now, we will use this option for our question-answer-generation (QAG).
2. Another option would be to use an existing llm with an specific prompt to extract questions and answers. As in https://github.com/amosproj/amos2024ss08-cloud-native-llm/blob/main/src/scripts/QA_enricher.py, one could design a prompt such as f"Write a question for the following input: '{input}'", where input corresponds to a text chunk with a predefined size from the gathered data. A possible problem that could arise is the model size overhead. An llm such as gemma-2b has 2.51B params (https://huggingface.co/google/gemma-2b), whereas the biggest qag model, t5-large, only has 738M params and t5-small only has 60.5M params. However it might be useful to try a qag scheme using an existing llm in order to improve our initial qa dataset.
3. Other resources that might be useful:
- https://github.com/dipta-dhar/Automatic-Question-Generator
- https://github.com/ramsrigouthamg/Questgen.ai



## Output format

There are 2 formats to fine tune a Q&A LLM:
1. Conversation format:
{"messages": [{"role": "system", "content": "You are..."}, {"role": "user", "content": "..."}, {"role": "assistant", "content": "..."}]}
2. Instruction format: 
{"prompt": "...prompt text...", "completion": "...generated text..."}

We should do more research on these 2 formats and see which one works best for our project, however the conversion between the 2 formats should be straightforward.


## Conclusion 
The lmqg package provides a straightforward and lightweight resource for generating both questions and answers given input text data. It will serve as a first approach for qag, later llms could be used for this task aswell as they might provide better results, but will be more resource intensive. For augmenting the question/answer dataset exisiting llms, such as GEMMA will be used, however lmqg might be feasible for this task aswell. Which resource to use will be determined by trial and testing the different resources.
